internal/proto/reader.go

```diff
@@ -92,240 +127,377 @@ func (r *Reader) readLine() ([]byte, error) {
+		n, err := replyLen(line)
 		if err != nil {
-			return nil, 0, err
+			return nil, err
 		}
-		keys[i] = key
+		return a(r, line[0], int64(n))
+	default:
+		return nil, fmt.Errorf("redis: can't parse aggregate(array/set/push/map) reply: %.100q", line)
 	}
-
-	return keys, cursor, err
 }
 
-func (r *Reader) ReadInt() (int64, error) {
-	b, err := r.readTmpBytesReply()
+func (r *Reader) ReadArrayReply(m MultiBulkParse) (interface{}, error) {
+	line, err := r.Pathfinder()
 	if err != nil {
-		return 0, err
+		return nil, err
+	}
+	switch line[0] {
+	case RespArray, RespSet, RespPush:
+		n, err := replyLen(line)
+		if err != nil {
+			return nil, err
+		}
+		return m(r, int64(n))
+	default:
+		return nil, fmt.Errorf("redis: can't parse array(array/set/push) reply: %.100q", line)
 	}
-	return util.ParseInt(b, 10, 64)
 }
 
-func (r *Reader) ReadUint() (uint64, error) {
-	b, err := r.readTmpBytesReply()
+func (r *Reader) ReadMapReply(m MultiBulkParse) (interface{}, error) {
+	line, err := r.Pathfinder()
 	if err != nil {
-		return 0, err
+		return nil, err
+	}
+	switch line[0] {
+	case RespMap:
+		n, err := replyLen(line)
+		if err != nil {
+			return nil, err
+		}
+		return m(r, int64(n))
+	case RespArray, RespSet, RespPush:
+		n, err := replyLen(line)
+		if err != nil {
+			return nil, err
+		}
+		return m(r, int64(n/2))
+	default:
+		return nil, fmt.Errorf("redis: can't parse map reply: %.100q", line)
 	}
-	return util.ParseUint(b, 10, 64)
 }
 
-func (r *Reader) ReadFloatReply() (float64, error) {
-	b, err := r.readTmpBytesReply()
+func (r *Reader) ReadArrayLen() (int, error) {
+	line, err := r.Pathfinder()
 	if err != nil {
 		return 0, err
 	}
-	return util.ParseFloat(b, 64)
+	switch line[0] {
+	case RespArray, RespSet, RespPush:
+		n, err := replyLen(line)
+		if err != nil {
+			return 0, err
+		}
+		return n, nil
+	default:
+		return 0, fmt.Errorf("redis: can't parse array(array/set/push) reply: %.100q", line)
+	}
 }
 
-func (r *Reader) readTmpBytesReply() ([]byte, error) {
-	line, err := r.ReadLine()
+func (r *Reader) ReadMapLen() (int, error) {
+	line, err := r.Pathfinder()
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 	switch line[0] {
-	case ErrorReply:
-		return nil, ParseErrorReply(line)
-	case StringReply:
-		return r._readTmpBytesReply(line)
-	case StatusReply:
-		return line[1:], nil
+	case RespMap:
+		n, err := replyLen(line)
+		if err != nil {
+			return 0, err
+		}
+		return n, nil
+	case RespArray, RespSet, RespPush:
+		n, err := replyLen(line)
+		if err != nil {
+			return 0, err
+		}
+		return n / 2, nil
 	default:
-		return nil, fmt.Errorf("redis: can't parse string reply: %.100q", line)
+		return 0, fmt.Errorf("redis: can't parse map reply: %.100q", line)
 	}
 }
 
-func (r *Reader) _readTmpBytesReply(line []byte) ([]byte, error) {
-	if isNilReply(line) {
-		return nil, Nil
+// Discard the data represented by line, if line is nil, read the next line.
+func (r *Reader) Discard(line []byte) (err error) {
```

Comment:
Let's remove the part that handles `if line is nil, read the next line.`. You can introduce a separate function like `DiscardNext()` if that is a common operation.